SAK-30601 : Upgraded basic LTI dashboard

frasese opened the pull request:
Upgraded basic LTI dashboard

ottenhoff commented:
Right now, it appears you need to set "basiclti.tool.date.format" to get a relevant date format for your area of the world.  The default format is not USA-friendly.

How about just using a nice default like ISO8601 style: yyyy-MM-dd'T'HH:mm:ssZ

frasese commented:
I've removed the dateformat parameter from sakai.properties and added it to the resource bundles. So it will change according to the user language.

Default dateformat will be yyyy-MM-dd'T'HH:mm:ssZ and en_GB dateformat will be dd/MM/yyyy HH:mm:ss

buckett commented:
Rather than requiring all the i18n people to set the date format could you not just use the standard JVM date formatters with code like:

    Locale userLocale = (new ResourceLoader()).getLocale();
    DateFormat formatter = DateFormat.getDateInstance(DateFormat.MEDIUM, userLocale);

ottenhoff commented:
Agree with @buckett . Thanks.

frasese commented:
Ok, I've done the changes but : 
- I've used "getDateTimeInstance" instead of "getDateInstance", so we can also show the added time.
- I've used the SHORT format instead of the MEDIUM one, because medium format contains a comma in the US format and can carry some problems with the CSV export.

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/basiclti/impl/BasicLTISecurityServiceImpl.java:
```diff
@@ -376,7 +381,58 @@ else if ( refId.startsWith("content:") && refId.length() > 8 )
+				} 
+				else if (refId.startsWith("export:") && refId.length() > 7) 
+				{
+		            final String[] tokens = refId.split(":");
```
Indentation doesn't look to align up here (tabs/spaces?)

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/LTIExportServiceImpl.java:
```diff
@@ -0,0 +1,254 @@
+	}
+	
+	private boolean isAdmin(String siteId) {
+		if ( siteId == null ) {
```
Why not just use `org.sakaiproject.authz.api.SecurityService#isSuperUser()`?

frasese replied:
Because is not checking for super user permissions just checks if has maintain access. This is copied from BaseLTIService, but maybe we should change the name for that method.

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/LTIExportServiceImpl.java:
```diff
@@ -0,0 +1,254 @@
+		
+		//set header row
+		result.append(rb.getString("export.title"));
+		result.append(getCSVSeparator());
```
This will break if a site title has a comma in it. When building a CSV you should correctly escape all the values, normally you do this by quoting all values in double quotes. Double quotes should be escaped with a preceding double quote. Although it's not complex you could use a CSV library.

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/LTIExportServiceImpl.java:
```diff
@@ -0,0 +1,254 @@
+	
+	private String getCSVSeparator() {
+		if(csvSeparator == null) {
+			csvSeparator = serverConfigurationService.getString("basiclti.export.csv.separator", ",");
```
Do you really need to change the deliminator in a CSV file?

frasese replied:
For example in Spain, default deliminator is ';'

buckett replied:
Ok, seems strange as if you're numbers include comma just quote them, but why would basiclti want to use something different from the sakai.property `csv.separator` which was introduced in https://jira.sakaiproject.org/browse/SAK-16405

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/LTIExportServiceImpl.java:
```diff
@@ -0,0 +1,254 @@
+				result.append(validator.escapeHtml((String)content.get("SITE_CONTACT_NAME"), false));
+				result.append(getCSVSeparator());
+				
+				result.append(validator.escapeHtml((String)content.get("SITE_CONTACT_EMAIL"), false));
```
This isn't being written to a HTML page so why should all the content be HTML escaped?

buckett replied:
When outputting CSVs you needs to escape special characters (comma) normally this is done by enclosing the field in double quotes and then escaping a double quote with an extra double quote.

Although this is simple there are small java libraries that will do this.

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/LTIExportServiceImpl.java:
```diff
@@ -0,0 +1,254 @@
+		for(Map<String,Object> content : contents) {
+			cellNum = 0;
+			row = sheet.createRow(rowNum++);
+			row.createCell(cellNum++).setCellValue(validator.escapeHtml((String)content.get("title"), false));
```
Again, why escape for HTML?

buckett commented on basiclti/basiclti-tool/src/java/org/sakaiproject/blti/tool/LTIAdminTool.java:
```diff
@@ -162,6 +194,152 @@ public String buildMainPanelContext(VelocityPortlet portlet, Context context,
+	 * Accepted parameters : id
+	 * @param data
+	 */
+	public void doFilter(RunData data) {
```
Doesn't match the code style of this class. Opening braces of new methods should be on their own line.

buckett commented on basiclti/basiclti-tool/src/java/org/sakaiproject/blti/tool/LTIAdminTool.java:
```diff
@@ -205,6 +434,30 @@ public String buildToolSitePanelContext(VelocityPortlet portlet, Context context
+		if (df instanceof SimpleDateFormat)
+		{
+			SimpleDateFormat sdf = (SimpleDateFormat)df;
+			context.put("dateToolFormat",sdf.toLocalizedPattern());
```
Why do you need the `dateToolFormat` in the context at all, when you create the formatter you specify the format and the locale so if you just call `df.format(Date)` it should give you the correct output shouldn't it?

frasese replied:
Yes, but I don't like to manually format dates in the java side. If we do that, we need to include in the context (or in this case, in the content object) the date and the formatted date. Now is just a field, but if in the future we add more date fields we will need to do this process for each one.

If we put the dateTool in the context, we can format the dates in the velocity file.

buckett replied:
Yep, but why not do:
```
context.put("dateTool", df);
```
then in the velocity file just do the formatting
```
$!dateTool.format($!content.get("created_at"))
```
The reason for using velocity DateTool class is so that Velocity does automatic formatting of dates without you having to explicitly say it should use a formatter, but at the moment this isn't being used so it's just as easy to pass in the Java date formatter.

frasese replied:
You are right, dateTool was useful at the beginning, but now it has no sense.

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/DBLTIService.java:
```diff
@@ -461,20 +463,48 @@ public Object updateContentDao(Long key, Object newProps, String siteId,
-			content.put("launch_url", getContentLaunch(content));
-		}
-		return contents;
+        String extraSelect = "SAKAI_SITE.TITLE AS SITE_TITLE, ssp1.VALUE AS SITE_CONTACT_NAME, ssp2.VALUE AS SITE_CONTACT_EMAIL";
```
Doing joins in the DB and not going through the services in Sakai is normally frowned upon as it means that any caching or refactoring of the original tables becomes much harder. It also means that security checks can be unintentionally bypassed.

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/basiclti/impl/BasicLTISecurityServiceImpl.java:
```diff
@@ -376,7 +381,58 @@ else if ( refId.startsWith("content:") && refId.length() > 8 )
+		                        }
+		                    }
+		                }
+		            }
```
What about handling the case where the export type isn't recognised?

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/basiclti/impl/BasicLTISecurityServiceImpl.java:
```diff
@@ -376,7 +381,58 @@ else if ( refId.startsWith("content:") && refId.length() > 8 )
+		                    out = (OutputStream)res.getOutputStream();
+		                    ltiExportService.export(out, ref.getContext(), exportType, filterId);
+		                    out.flush();
+		                    out.close();
```
You don't need this as it will be done in the finally section.

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/basiclti/impl/BasicLTISecurityServiceImpl.java:
```diff
@@ -376,7 +381,58 @@ else if ( refId.startsWith("content:") && refId.length() > 8 )
+		                    logger.warn(": lti export " + e.getMessage());
+		                    if (out != null) {
+		                        try {
+		                            out.close();
```
You don't need this close as it will be done in the finally section.

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/LTIExportServiceImpl.java:
```diff
@@ -0,0 +1,254 @@
+					url = (String)(ltiService.getToolDao(new Long(content.get(LTIService.LTI_TOOL_ID).toString()), siteId).get("launch"));
+				} catch(Exception e) {
+					url = "-";
+					e.printStackTrace();
```
Shouldn't have `e.printStackTrace()` use logging instead, although this seems to be expected so does it need logging at all?

buckett commented on basiclti/basiclti-api/src/java/org/sakaiproject/lti/api/LTISearchData.java:
```diff
@@ -0,0 +1,30 @@
+		return (search != null);
+	}
+	
+	public void addSerchValue(Object value) {
```
Typo in method name, should be `addSearchValue`.

buckett commented on basiclti/basiclti-api/src/java/org/sakaiproject/lti/api/LTISearchData.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class LTISearchData {
```
JavaDoc on class just explaining what it's for.

buckett commented on basiclti/basiclti-api/src/java/org/sakaiproject/lti/api/LTIService.java:
```diff
@@ -884,5 +916,10 @@
+    public static final String LTI_SEARCH_TOKEN_SEPARATOR = "#:#";
+    public static final String LTI_SITE_ATTRIBUTION_PROPERTY_KEY = "basiclti.tool.site.attribution.key";
+    public static final String LTI_SITE_ATTRIBUTION_PROPERTY_NAME = "basiclti.tool.site.attribution.name";
+    public static final String LTI_DATE_FORMAT = "basiclti.tool.date.format";
```
This constant is unused.

buckett commented on basiclti/basiclti-impl/src/java/org/sakaiproject/lti/impl/LTIExportServiceImpl.java:
```diff
@@ -0,0 +1,254 @@
+		return csvSeparator;
+	}
+	
+	private void exportCSV(OutputStream out, String siteId, List<Map<String,Object>> contents) {
```
Would have been nicer to have simple abstraction of XLS/CSV so that the logic about what to display isn't duplicated but too late now.

buckett commented on basiclti/basiclti-tool/src/webapp/vm/lti_tool_site.vm:
```diff
@@ -121,52 +104,100 @@ ${includeLatestJQuery}
 			</td>
-			#if ($isAdmin) <td headers="site"> $validator.escapeHtml($!content.get("SITE_ID"))</a></td> #end
+			#if ($isAdmin) 
+				<td headers="site"><a href="/portal/site/$validator.escapeHtml($!content.get("SITE_ID"))">$validator.escapeHtml($!content.get("SITE_TITLE"))</a></td>
```
You should use `Site.getUrl()` as this will correctly build URLs to sites.

buckett commented on basiclti/basiclti-tool/src/webapp/vm/lti_tool_site.vm:
```diff
@@ -121,52 +104,100 @@ ${includeLatestJQuery}
+		#if ($contents && $contents.size() > 0)
+		<div class="navPanel">
+			<div class="listNav">
+				<a style="float:right" href="$export_url_excel"><input type="submit" value="$tlang.getString("content.export.excel")" /></a>
```
Should be hidden from non-admins if the actual exports don't allow non-admins.

buckett commented on basiclti/basiclti-api/src/java/org/sakaiproject/lti/api/LTIExportService.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface LTIExportService {
+
+	public static final int TYPE_CSV = 0;
```
Would be nice to use `enum` for better type safety.

buckett commented on basiclti/basiclti-api/src/java/org/sakaiproject/lti/api/LTIService.java:
```diff
@@ -201,6 +201,15 @@
+	 * @param exportType
+	 * @return
+	 */
+	public String getExportLaunch(String siteId, String filterId, int exportType);
```
Method name isn't very clear, as this returns an export URL not a URL that starts an LTI launch. Something like `getExportUrl` would be much clearer.

buckett commented:
@jonespm @ottenhoff @ern It looks like basic-lti is doing joins to SAKAI_SITE table directly, generally this has been frowned on in the past because of:

- Limits DB refactoring on the original tables.
- Permissions checks are bypassed.
- Caches are bypassed.
- Cache invalidation becomes trick.

Are you happy with these further changes that bind basic-lti directly the site tables?

frasese commented:
- Implemented all changes proposed by @buckett 
- Replaced search field by a selector in the "attribution" dynamic column

I've let the changes in an additional commit, so you can track them better. I will squash them when you give me the ok.

frasese commented:
Added a second update.
Again,  if everything is ok, I will squash all commits.

ottenhoff commented:
/CC: @buckett  and @csev any comments? Do you want to merge?

ottenhoff commented:
Closing this PR because it is against 11.x
